### PR TITLE
LPS-105721 Make `useIsMounted` work in child component edge case

### PR DESCRIPTION
**Pull request by wincent**
**tl;dr:** this fixes an obscure edge case that we ran into in the React content page editor.

For anybody curious about React inner workings, full description follows:

---

This commit makes the `useIsMounted` hook work from inside child components' on-mount `useEffect` callbacks. It's very edge-casey, but is something we just saw in the React content page editor when we tried to move from `useState` for managing the currently active sidebar panel, to using our reducer-based store.

Our top-level `useReducer` hook uses `useIsMounted` internally (via `useThunk`) to prevent us from ever accidentally dispatching actions when we have unmounted, but we found an edge case in the sidebar where actions were not getting propagated during mount. The sequence basically looks like this:

1. Top-level app component renders (sets up reducer).
2. Grand-child component (`<Sidebar />`) does initial render.
3. Grand-child does a `useEffect` to run a side effect (loading the first panel into view; note that because this app is dynamic we don't actually know what the first panel will be until we run).

That last step worked fine when we using `useState` to set the active panel ID, but stopped working when we switched to `dispatch`. The reason is that the `dispatch` method comes from the top-level, and it is wrapped in a call to `useThunk` which internally uses `useIsMounted`. `useIsMounted` uses `useEffect` to toggle its internal "is-mounted" book-keeping from `false` to `true` once it has mounted.

Thus, when the grand-child invokes `dispatch` from inside `useEffect`, it is relying on the value returned by `isMounted()` as determined by the top-level component. But child component `useEffect` callbacks run before their parents', so `isMounted()` actually ends up returning `false`. In general, in React, "being mounted" means "component and all of its descendants are mounted", so children get notified first.

The fix is to make `useIsMounted` update its book-keeping a little earlier, before the child component `useEffect` runs. We do this by switching to `useLayoutEffect`. This is the execution order:

1. Child `useLayoutEffect` runs.
2. Parent `useLayoutEffect` runs.
3. Child `useEffect` runs.
4. Parent `useEffect` runs.

In other words, `useIsMounted` is now more robust for the common use cases (which all use `useEffect`). The rarer use cases (using `useLayoutEffect`) are still subject to this quirk of children running their callbacks before their parents. Most of the time we use `useIsMounted` locally and there is no reason to worry about any of this, but in a pattern that is probably going to become common in the future — using `useReducer` in conjunction with `useThunk` — the change in this commit makes us more robust.

### CODE REVIEW

**Comment by wincent** [comment]
cc @julien @mateomustapic

**Comment by brianchandotcom** [comment]
Merged. Thank you.
View total diff: https://github.com/brianchandotcom/liferay-portal/compare/af2934c0a2...352eafa1ab
